### PR TITLE
Provide option for tab to apply completion

### PR DESCRIPTION
**Pull request by mskar**
In #12575, I requested a feature that adds parentheses after functions completed with `Tab` or `Enter`. 
This pull request implements the feature I requested (function completion with added parentheses), but also strives to make autocompletion work like in the VSCode Python extension and PyCharm. In VSCode and PyCharm, `Tab` applies completion instead of cycling through completion options, like in bash or zsh.

This feature is enabled by default, but it can be disabled by adding 
```python
c.InteractiveShell.tab_apply_completion = False
```
to `ipython_config.py`. 

For more information, please take a look at [this comment](https://github.com/ipython/ipython/issues/12575#issuecomment-702484214) in #12575.

To install IPython with the `tab_apply_completion` feature, run 
`python -m pip install git+https://github.com/mskar/ipython.git@tab_apply_completion`
## TLDR: 
previous behavior: 
- `Tab` cycles completion
- function completion does not add parentheses 

proposed behavior: 
- `Tab` applies completion
- function completion adds parentheses

## Changes:
Changed: when triggering completion with `Tab`, insert text that all options have in common
Changed: use `Tab` to accept only option or accept selected
Changed: do not move through options with not `Tab` and `Shift-Tab`
Unchanged: use `Enter` to accept if completion menu is showing
Unchanged: move through options with `c-n` and `c-p`
Unchanged: cancel completion with `c-g`

### CODE REVIEW

**Comment by mskar** [comment]
I split the original feature into 5 independent features.
Each feature has its own setting (disabled by default) to allow for granular control of completion behavior:

1. `auto_complete_function_parentheses`:
  - Tab: Unchanged unless combined with one of the settings that end in `_option_on_tab`.
  - Enter: Accepts selected completion and add parentheses to functions
2. `auto_complete_selected_option_on_tab`:
  - Tab: Accepts the selected completion option if an option is selected
  - Enter: Unchanged
3. `auto_complete_top_option_on_enter`:
  - Tab: Unchanged
  - Enter: Accepts the first completion option if the completion menu is showing
4. `auto_complete_top_option_on_tab`:
  - Tab: Accepts the first completion option if the completion menu is showing and no option is selected
  - Enter: Unchanged
5. `auto_complete_only_option_on_tab`:
  - Tab: Accepts completion if there is only one completion option or starts completion if there are multiple options
  - Enter: Unchanged

These features can be combined.
Enabling all of the features makes completion behave like in PyCharm and VSCode.

Is 5 settings too many?
Would it be a good idea to add another setting that combined 2-5 (`auto_complete_[selected,top,only]_option_on_[tab,enter]`)?

Please let me know if any of the settings should be renamed.

**Comment by Carreau** [comment]
Thanks for your patience and sending this; I was not able to keep up with everything and to keep the number of PRs small I'm going to close this one; If you believe this was a mistake, feel free to comment or re-send it; 

I can't guarantee I'll get to it as I do not have the bandwidth.
